### PR TITLE
document const generics

### PR DESCRIPTION
**Pull request by lcnr**
blocked on the stabilization of `min_const_generics`.

### CODE REVIEW

**Comment by shirshak55** [comment]
Its merged in https://github.com/rust-lang/rust/pull/79135

File: src/const_eval.md [hunk 1]
```diff
@@ -112,6 +113,7 @@ Conversely, the following are possible in a const function, but not in a const c
 [comparison]:           expressions/operator-expr.md#comparison-operators
 [const functions]:      items/functions.md#const-functions
 [constants]:            items/constant-items.md
+[Const parameters]:    items/generics.md
```
Comment (est31) [inline src/const_eval.md:116]:
```suggestion
[Const parameters]:     items/generics.md
```

File: src/items/generics.md [hunk 2]
```diff
@@ -18,20 +19,34 @@
...
 Functions, type aliases, structs, enumerations, unions, traits, and
-implementations may be *parameterized* by types and lifetimes. These parameters
-are listed in angle <span class="parenthetical">brackets (`<...>`)</span>,
+implementations may be *parameterized* by types, constants and lifetimes. These
+parameters are listed in angle <span class="parenthetical">brackets (`<...>`)</span>,
 usually immediately after the name of the item and before its definition. For
 implementations, which don't have a name, they come directly after `impl`.
-Lifetime parameters must be declared before type parameters. Some examples of
-items with type and lifetime parameters:
+The order of generic parameters is restricted to lifetime parameters, then type parameters and then const parameters.
```
Comment (ehuss) [inline src/items/generics.md:35]:
```suggestion
The order of generic parameters is restricted to lifetime parameters, then type parameters, and then const parameters.
```

File: src/items/generics.md [hunk 3]
```diff
@@ -18,20 +19,34 @@
...
+> _ConstParams_:\
+> &nbsp;&nbsp; ( _ConstParam_ `,` )<sup>\*</sup> _ConstParam_<sup>?</sup>
+>
+> _ConstParam_:\
+> &nbsp;&nbsp; [_OuterAttribute_]<sup>?</sup> `const` [IDENTIFIER] `:` [_Type_]
 
 Functions, type aliases, structs, enumerations, unions, traits, and
-implementations may be *parameterized* by types and lifetimes. These parameters
-are listed in angle <span class="parenthetical">brackets (`<...>`)</span>,
+implementations may be *parameterized* by types, constants and lifetimes. These
```
Comment (ehuss) [inline src/items/generics.md:31]:
```suggestion
implementations may be *parameterized* by types, constants, and lifetimes. These
```

File: src/items/generics.md [hunk 4]
```diff
@@ -18,20 +19,34 @@
...
-items with type and lifetime parameters:
+The order of generic parameters is restricted to lifetime parameters, then type parameters and then const parameters.
+
+The only allowed types of const parameters are `u8`, `u16`, `u32`, `u64`, `u128`, `usize`
+`i8`, `i16`, `i32`, `i64`, `i128`, `isize`, `char` and `bool`.
+Const parameters may only be be used as standalone arguments inside
+of [types] and [repeat expressions].
+They can be used freely outside of [const contexts].
+
+Some examples of items with type, const and lifetime parameters:
```
Comment (ehuss) [inline src/items/generics.md:43]:
```suggestion
Some examples of items with type, const, and lifetime parameters:
```

File: src/items/generics.md [hunk 5]
```diff
@@ -18,20 +19,34 @@
...
 implementations, which don't have a name, they come directly after `impl`.
-Lifetime parameters must be declared before type parameters. Some examples of
-items with type and lifetime parameters:
+The order of generic parameters is restricted to lifetime parameters, then type parameters and then const parameters.
+
+The only allowed types of const parameters are `u8`, `u16`, `u32`, `u64`, `u128`, `usize`
+`i8`, `i16`, `i32`, `i64`, `i128`, `isize`, `char` and `bool`.
+Const parameters may only be be used as standalone arguments inside
+of [types] and [repeat expressions].
+They can be used freely outside of [const contexts].
```
Comment (ehuss) [inline src/items/generics.md:41]: I find these sentences a little difficult to understand.  Can you add more detail, along with some examples?

**Review by ehuss — COMMENTED** [review]
Maybe I'm confused, but it seems like the grammar for _GenericArgs_ needs to be updated for expressions?

Also, overall I feel like this is missing a lot of content.  There seems to be a large amount of information in the stabilization report that would be good to carry over.  There are a lot of details that seem to be missing (like const parameters don't have to be used unlike types and lifetimes).  I also see a number of things in the RFC that don't seem to be mentioned.

I appreciate the effort for adding this documentation, but I would like to see a bit more detail if possible.

File: src/items/generics.md [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
 
 > **<sup>Syntax</sup>**\
 > _Generics_ :\
 > &nbsp;&nbsp; `<` _GenericParams_ `>`
 >
 > _GenericParams_ :\
 > &nbsp;&nbsp; &nbsp;&nbsp; _LifetimeParams_\
-> &nbsp;&nbsp; | ( _LifetimeParam_ `,` )<sup>\*</sup> _TypeParams_
+> &nbsp;&nbsp; | ( _LifetimeParam_ `,` )<sup>\*</sup> _TypeParams_\
+> &nbsp;&nbsp; | ( _LifetimeParam_ `,` )<sup>\*</sup> ( _TypeParam_ `,` )<sup>\*</sup> _ConstParams_
```
[the remark this answers is not in the document]
Comment (lcnr) [reply src/items/generics.md:10]: won't do this in that PR, as I don't know what it's expected to look like now.

File: src/items/generics.md (hunk 5 above)
Comment (lcnr) [reply src/items/generics.md]: extended this a bit, it feels weird to go more in depth here as the explanation for types and lifetimes are also fairly short

**Comment by lcnr** [comment]
there is still a lot we could add about const generics or generics in general.

I am not able to spend a lot of time on rust in the near future so I would prefer merging this without adding too much more

File: src/paths.md [hunk 7]
```diff
@@ -52,18 +52,26 @@ mod m {
...
+> &nbsp;&nbsp; | `<` _GenericArgsLifetimes_ `,` _GenericArgsTypes_ `,` _GenericArgsConsts_ `,` _GenericArgsBindings_ `,`<sup>?</sup> `>`
 >
 > _GenericArgsLifetimes_ :\
 > &nbsp;&nbsp; [_Lifetime_] (`,` [_Lifetime_])<sup>\*</sup>
 >
 > _GenericArgsTypes_ :\
 > &nbsp;&nbsp; [_Type_] (`,` [_Type_])<sup>\*</sup>
 >
+> _GenericArgsConsts_ :\
+> &nbsp;&nbsp; [_Expression_] (`,` [_Expression_])<sup>\*</sup>
```
Comment (ehuss) [inline src/paths.md:73]: I don't think this is the actual grammar for arguments is it?  According to [`parse_generic_arg`](https://github.com/rust-lang/rust/blob/bbcaed03bf5505f3fed351887769ed1531599502/compiler/rustc_parse/src/parser/path.rs#L525-L537) I believe it is something like this:

````diff
diff --git a/src/paths.md b/src/paths.md
index 9126e5b..47a14ee 100644
--- a/src/paths.md
+++ b/src/paths.md
@@ -70,7 +70,13 @@ mod m {
 > &nbsp;&nbsp; [_Type_] (`,` [_Type_])<sup>\*</sup>
 >
 > _GenericArgsConsts_ :\
-> &nbsp;&nbsp; [_Expression_] (`,` [_Expression_])<sup>\*</sup>
+> &nbsp;&nbsp; _GenericArgsConst_ (`,` _GenericArgsConst_)<sup>\*</sup>
+>
+> _GenericArgsConst_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [_BlockExpression_]\
+> &nbsp;&nbsp; | [_LiteralExpression_]\
+> &nbsp;&nbsp; | `-` [_LiteralExpression_]\
+> &nbsp;&nbsp; | [_SimplePathSegment_]
 >
 > _GenericArgsBindings_ :\
 > &nbsp;&nbsp; _GenericArgsBinding_ (`,` _GenericArgsBinding_)<sup>\*</sup>
@@ -375,10 +381,13 @@ mod without { // ::without
 # fn main() {}
 ```

+[_BlockExpression_]: expressions/block-expr.md
+[_Expression_]: expressions.md
 [_GenericArgs_]: #paths-in-expressions
 [_Lifetime_]: trait-bounds.md
+[_LiteralExpression_]: expressions/literal-expr.md
+[_SimplePathSegment_]: #simple-paths
 [_Type_]: types.md#type-expressions
-[_Expression_]: expressions.md
 [literal]: expressions/literal-expr.md
 [item]: items.md
 [variable]: variables.md
````

Is that correct?

**Comment by ehuss** [comment]
> I am not able to spend a lot of time on rust in the near future so I would prefer merging this without adding too much more

No worries, if you can confirm the comment above about the grammar, I can probably merge this and work on extending it.

File: src/paths.md (hunk 7 above)
Comment (lcnr) [reply src/paths.md:73]: yeah, that looks right to me

File: src/paths.md [hunk 8]
```diff
@@ -81,6 +89,9 @@ ambiguity with the less-than operator. This is colloquially known as "turbofish"
 Vec::<u8>::with_capacity(1024);
 ```
 
+Const arguments must be surrounded by braces unless they are a
+[literal] or a single segment path.
```
Comment (lcnr) [inline src/paths.md:93]: in this case this section might not be needed

**Review by ehuss — APPROVED** [review]
I went ahead and pushed the fix for the grammar.  I'll follow up with other PRs.

**Pull request merged**